packages/shared/src/utils.ts

```diff
@@ -40,6 +40,16 @@ export function toArray(arr: NdArray<number[]> | number[]): number[] {
   return 'data' in arr ? arr.data : arr;
 }
 
+export function toTypedNdArray<T extends number[] | TypedArray>(
```

Comment:
Would it be possible for the typing to return `NdArray<Float32Array>` when `T extends number[]` ?

So that we know that `toTypedNdArray` does not convert to _any_ TypedArray ?

Reply:
Re-reading your comment, I'd like to clarify that the return type `NdArray<Exclude<T, number[]> | Float32Array>`  does not mean _any_ TypedArray. It keeps whichever TypeArrays were allowed in the input parameter. So when we call it in `HeatmapVis` with `NdArray<number[] | CompatibleTypedArray>`, we get back `NdArray<CompatibleTypedArray>` (since `CompatibleTypedArray` already includes `Float32Array`).

The only downside of the current TS signature is that if you call the function with an ndarray of type `NdArray<Uint16Array>`, you'll get back `NdArray<Uint16Array | Float32Array>`.

But the function is not really meant to be called with an already typed ndarray -- it's meant to be used when you don't know whether you have a typed of `number[]` ndarray like in the `HeatmapVis`.

Reply:
> Re-reading your comment, I'd like to clarify that the return type `NdArray<Exclude<T, number[]> | Float32Array>` does not mean _any_ TypedArray. It keeps whichever TypedArrays were allowed in the input parameter. So when we call it in `HeatmapVis` with `NdArray<number[] | CompatibleTypedArray>`, we get back `NdArray<CompatibleTypedArray>` (since `CompatibleTypedArray` already includes `Float32Array`).

Understood, but I thought we could make the return type stronger. You could have `T extends number[] ? NdArray<Float32Array> : NdArray<Exclude<T, number[]>` to avoid the issue you mention below.
> 
> The only downside of the current TS signature is that if you call the function with an ndarray of type `NdArray<Uint16Array>`, you'll get back `NdArray<Uint16Array | Float32Array>`.

Reply:
Doesn't work, unfortunately. I get errors on the two return statements saying that the type of what I'm returning is not assignable to the function's return conditional type without further explanations...

Reply:
Yeah, I figured that it would be a bit too tortuous for TypeScript to make its way. Let's keep it like this then